src/audio/copier.c

```diff
@@ -169,71 +191,149 @@ static struct comp_dev *create_host(struct comp_dev *parent_dev, struct copier_d
 
 	drv = ipc4_get_drv((uint8_t *)&host);
 	if (!drv)
-		return NULL;
+		return -EINVAL;
 
 	parent_dev->ipc_config.type = SOF_COMP_HOST;
 	config->type = SOF_COMP_HOST;
 
-	create_endpoint_buffer(parent_dev, cd, config, copier_cfg);
+	cd->endpoint = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			       sizeof(struct comp_dev *));
+	if (!cd->endpoint)
+		return -ENOMEM;
+
+	cd->endpoint_buffer = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+				      sizeof(struct comp_buffer *));
+	if (!cd->endpoint_buffer) {
+		rfree(cd->endpoint);
+		return -ENOMEM;
+	}
+
+	create_endpoint_buffer(parent_dev, cd, config, copier_cfg, ipc4_gtw_host, 0);
 
 	memset(&ipc_host, 0, sizeof(ipc_host));
 	ipc_host.direction = dir;
 
 	dev = drv->ops.create(drv, config, &ipc_host);
 	if (!dev)
-		return NULL;
+		return -EINVAL;
 
 	list_init(&dev->bsource_list);
 	list_init(&dev->bsink_list);
 
 	if (cd->direction == SOF_IPC_STREAM_PLAYBACK) {
-		comp_buffer_connect(dev, config->core, cd->endpoint_buffer,
+		comp_buffer_connect(dev, config->core, cd->endpoint_buffer[cd->endpoint_num],
```

Comment:
sure

Reply:
and I still don't see it fixed?

Reply:
I didn't get your idea. It is a pointer array and allocated dynamically.

Reply:
@lyakh can you give a C example of your change. Thanks !

Reply:
@lgirdwood @RanderWang I'm referring to my comment to `rzalloc()` in line 199 above:
```
+	cd->endpoint = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			       sizeof(struct comp_dev *));
```
That allocates a single pointer, 4 bytes. This is a `SHARED` allocation so one would assume, that it doesn't have to be cache-line aligned, but with Zephyr all allocations are currently cache-line (64-byte) aligned. In line 205 a similar allocation is performed. So that creates quite a lot of wasted SRAM. Also, on Zephyr each allocation adds a 4-byte header, possibly in yet another cache line. I understand, that `.endpoint` and `.endpoint_buffer` should be arrays of pointers, and since there's always just one host we allocate one pointer, but that seems like a lot of wasted RAM. And we do run into `-ENOMEM` from time to time even on newer cAVS 2.5 and similar systems, don't we? The first trivial optimisation could be placing `endpoint` and `endpoint_buffer` into a structure since there are always an equal number of both and allocating only once. A slightly more complex change, but still possibly doable would be counting DAIs before allocating the copier data object and then allocating it all in one go. We do a LOT of allocations, and we're running on a DSP with 100s of kilobytes to low megabytes of SRAM...

Reply:
Thanks,  I fully agree with you.  Actually the max number of endpoint is 4 and I wanted to save 2 or 3 pointer memory of 2 to 3 * 32bits then introduced dynamic allocation. It is a bad deal. Now I use static array of 4 and  waste less than 3 * 32bits size memory.

Reply:
sure, I didn't use static data just fixed size of array. My explanation was confusing.